File: src/assets/scripts/generateFolder.js
import path from "path";
import fs from "fs";
import pkg from 'canvas';
const { createCanvas, registerFont } = pkg;
import { Canvg } from "canvg";
import pkg2 from 'jsdom';
const { JSDOM } = pkg2;
import pngToIco from "png-to-ico";

import config from "../../../.OLD/backend/config/index.ts";
import recolor from "./recolor.js";
import injectFont from "./injectFont.js";
import { snowflake } from "../../../.OLD/backend/server.ts";
import { getColor, scaleSVG } from "../../../.OLD/backend/helpers/index.ts";

async function generateFolder(options = {}) {
    let {
        style = "shaded",
        variant = "left1",
        baseColor = "#FFD65C",
        backColor = "#000000",
        iconColor = "#000000",
        mediumIcon = "",
        smallIcon = "",
        text = "",
        saturation = 1,
        brightness = 1,
        contrast = 1,
        isCustomBackColor = false,
        isCustomIconColor = false,
        folder = "",
        saveSVG = false,
        savePNG = false,
        saveICO = true,
        width = 256,
        height = 256,
    } = options;

    baseColor = getColor(baseColor);
    backColor = getColor(backColor);
    iconColor = getColor(iconColor);

    const fontDir = path.join(config.folders.assets, "fonts", "jetbrains");
    const svgDir = path.join(config.folders.assets, "svg", style);

    [config.folders.assets, fontDir, svgDir, config.folders.generated].forEach((p) => {
        if (!fs.existsSync(p)) fs.mkdirSync(p, { recursive: true });
    });

    const fontPath = path.join(fontDir, "nerdfont.ttf");
    if (fs.existsSync(fontPath)) {
        registerFont(fontPath, { family: "jetbrains-nerdfont" });
    } else {
        console.warn(`Font file not found: ${fontPath}. Text rendering may fail.`);
    }

    const svgPath = path.join(svgDir, `${variant}.svg`);
    if (!fs.existsSync(svgPath)) throw new Error(`SVG template not found: ${svgPath}`);
    let svg = fs.readFileSync(svgPath, "utf8");

    svg = recolor(
        svg,
        style,
        variant,
        baseColor,
        backColor,
        iconColor,
        mediumIcon,
        smallIcon,
        text,
        saturation,
        brightness,
        contrast,
        isCustomBackColor,
        isCustomIconColor
    );

    if (mediumIcon || smallIcon || text) {
        svg = injectFont(svg);
    }

    const genFolder = path.join(config.folders.generated, style, variant, folder);
    fs.mkdirSync(genFolder, { recursive: true });

    let id = "";
    if (saveSVG || savePNG || saveICO) {
        id = snowflake.generate();
    }

    if (saveSVG) {
        const svgFile = path.join(genFolder, `${id}.svg`);
        fs.writeFileSync(svgFile, svg, "utf8");
    }

    const { window } = new JSDOM(`<!DOCTYPE html>`);
    const DOMParser = window.DOMParser;

    let pngBuffer = null;
    if (savePNG) {
        const canvas = createCanvas(width, height);
        const ctx = canvas.getContext("2d");

        const scaledSVG = scaleSVG(svg, width, height);
        const v = Canvg.fromString(ctx, scaledSVG, { DOMParser });
        await v.render();

        pngBuffer = canvas.toBuffer("image/png");
        const pngFile = path.join(genFolder, `${id}.png`);
        fs.writeFileSync(pngFile, pngBuffer);
    }

    if (saveICO) {
        const sizes = [16, 32, 48, 64, 128, 256];
        const pngBuffers = [];

        for (const size of sizes) {
            const canvas = createCanvas(size, size);
            const ctx = canvas.getContext("2d");
            ctx.imageSmoothingQuality = "high";

            const scaledSVG = scaleSVG(svg, size, size);
            const v = Canvg.fromString(ctx, scaledSVG, { DOMParser, ignoreDimensions: true });
            await v.render();

            pngBuffers.push(canvas.toBuffer("image/png"));
        }

        try {
            const icoBuffer = await pngToIco(pngBuffers);
            const icoFile = path.join(genFolder, `${id}.ico`);
            fs.writeFileSync(icoFile, icoBuffer);
        } catch (err) {
            console.error("Failed to generate ICO:", err);
        }
    }

    return { id, options: {
            style,
            variant,
            baseColor,
            backColor,
            iconColor,
            mediumIcon,
            smallIcon,
            text,
            saturation,
            brightness,
            contrast,
            isCustomBackColor,
            isCustomIconColor,
            folder,
            saveSVG,
            savePNG,
            saveICO,
            width,
            height,
        }
    }
}

export default generateFolder;

File: src/cornelia/scripts/folders/generate_folder.py
from io import BytesIO
from pathlib import Path
from PIL import Image
from rich import print

from src.cornelia.scripts.config import config
from src.cornelia.scripts.folders.recolor import recolor
from src.cornelia.scripts.folders.inject_font import inject_font
from src.cornelia.scripts.folders.svg_to_png import svg_to_png
from src.cornelia.scripts.helpers.snowflake import generate_snowflake
from src.cornelia.scripts.folders.color_management import get_color
from src.cornelia.scripts.folders.scale_svg import scale_svg

async def generate_folder(options: dict = None):
    if options is None:
        options = {}

    style = options.get("style", "papirus")
    variant = options.get("variant", "left1")
    base_color = get_color(options.get("baseColor", "#16a085"))
    back_color = get_color(options.get("backColor", "#12806a"))
    icon_color = get_color(options.get("iconColor", "#08382e"))
    medium_icon = options.get("mediumIcon", "")
    small_icon = options.get("smallIcon", "")
    text = options.get("text", "")
    saturation = options.get("saturation", 1)
    brightness = options.get("brightness", 1)
    contrast = options.get("contrast", 1)
    is_custom_back_color = options.get("isCustomBackColor", False)
    is_custom_icon_color = options.get("isCustomIconColor", False)
    folder = options.get("folder", "")
    save_svg = options.get("saveSVG", True)
    save_png = options.get("savePNG", False)
    save_ico = options.get("saveICO", False)
    width = options.get("width", 256)
    height = options.get("height", 256)

    font_dir = Path(config["folders"]["root"]) / "src" / "userdata" / "packs" / "official" / "assets" / "fonts"
    svg_dir = Path(config["folders"]["root"]) / "src" / "userdata" / "packs" / "official" / "assets" / "folders" / style
    gen_folder = Path(config["folders"]["generated"]) / style / variant / folder

    for p in [font_dir, svg_dir, gen_folder]:
        p.mkdir(parents=True, exist_ok=True)
        
    svg_path = svg_dir / f"{variant}.svg"

    if not svg_path.exists():
        raise Exception(f"[red]󰜡 SVG template not found: {svg_path}[/red]")

    svg = svg_path.read_text(encoding="utf-8")

    svg = recolor(
        svg,
        style,
        variant,
        base_color,
        back_color,
        icon_color,
        medium_icon,
        small_icon,
        text,
        saturation,
        brightness,
        contrast,
        is_custom_back_color,
        is_custom_icon_color,
    )

    if medium_icon or small_icon or text:
        svg = inject_font(svg)

    file_id = None
    if save_svg or save_png or save_ico:
        file_id = generate_snowflake()

    """
    ————————————————————————————————————————————————————————————————
    SVG Output
    ———————————————————————————————————————————————————————————————— 
    """

    if save_svg and file_id:
        (gen_folder / f"{file_id}.svg").write_text(svg, encoding="utf-8")

    """
    ————————————————————————————————————————————————————————————————
    PNG Output
    ———————————————————————————————————————————————————————————————— 
    """

    image = None

    if save_png or save_ico:
        scaled_svg = scale_svg(svg, width, height)
        png_bytes = svg_to_png(scaled_svg, width, height)
        image = Image.open(BytesIO(png_bytes)).convert("RGBA")

        if save_png and file_id:
            image.save(gen_folder / f"{file_id}.png")

    """
    ————————————————————————————————————————————————————————————————
    ICO Output
    ———————————————————————————————————————————————————————————————— 
    """

    if save_ico and file_id:
        sizes = [(s, s) for s in [16, 32, 48, 64, 128, 256]]

        try:
            scaled_svg = scale_svg(svg, 256, 256)
            png_bytes = svg_to_png(scaled_svg, 256, 256)

            base_img = Image.open(BytesIO(png_bytes)).convert("RGBA")

            ico_path = gen_folder / f"{file_id}.ico"

            base_img.save(
                ico_path,
                format="ICO",
                sizes=sizes
            )

        except Exception as e:
            print("Failed to generate ICO:", e)

    """
    ————————————————————————————————————————————————————————————————
    Return
    ———————————————————————————————————————————————————————————————— 
    """

    return {
        "id": file_id,
        "path": str(gen_folder),
        "options": {
            "style": style,
            "variant": variant,
            "baseColor": base_color,
            "backColor": back_color,
            "iconColor": icon_color,
            "mediumIcon": medium_icon,
            "smallIcon": small_icon,
            "text": text,
            "saturation": saturation,
            "brightness": brightness,
            "contrast": contrast,
            "isCustomBackColor": is_custom_back_color,
            "isCustomIconColor": is_custom_icon_color,
            "folder": folder,
            "saveSVG": save_svg,
            "savePNG": save_png,
            "saveICO": save_ico,
            "width": width,
            "height": height,
        },
    }